src/pipeline/src/etl/transform/transformer/greptime.rs

```diff
@@ -431,11 +437,72 @@ pub fn identity_pipeline(
     }
 }
 
+/// Consumes the JSON object and consumes it into a single-level object.
+///
+/// The `max_nested_levels` parameter is used to limit the nested levels of the JSON object.
+/// The error will be returned if the nested levels is greater than the `max_nested_levels`.
+pub fn flatten_json_object(
+    object: Map<String, JsonValue>,
+    max_nested_levels: usize,
+) -> Result<Map<String, JsonValue>> {
+    let mut flattened = Map::new();
+
+    if !object.is_empty() {
+        // it will use recursion to flatten the object.
+        do_flatten_json_object(&mut flattened, None, object, 1, max_nested_levels)?;
+    }
+
+    Ok(flattened)
+}
+
+fn do_flatten_json_object(
+    dest: &mut Map<String, JsonValue>,
+    base: Option<&str>,
+    object: Map<String, JsonValue>,
+    current_level: usize,
+    max_nested_levels: usize,
+) -> Result<()> {
+    // For safety, we do not allow the depth to be greater than the max_object_depth.
+    if current_level > max_nested_levels {
+        return ReachedMaxNestedLevelsSnafu { max_nested_levels }.fail();
+    }
+
+    for (key, value) in object {
+        let new_key = base.map_or_else(|| key.clone(), |base_key| format!("{base_key}.{key}"));
```

Comment:
Does the `key.clone()` look unnecessary? We can consume it.

Reply:
`key.clone()` occurs when the `base` is `None`, which is during the initial call, and in this case, the `clone()` is necessary.